### ISSUE TITLE
Hypothesis fixture scope warning does not detect if fixture is overridden

### ISSUE THREAD

**Issue by simonfagerholm** [HypothesisWorks/hypothesis#2375]
In #2356 a warning was added if `@given()` is used together with function scoped fixtures.
However the warning functionality does not discover if a fixture is overridden by a "closer" definition, e.g. if a fixture with the same name is defined in the test file. See image below of how `fx_defs` looks with multiple definitions.
![image](https://user-images.githubusercontent.com/21139421/77293095-edb93200-6ce1-11ea-8222-b98dc9090748.png)
This was discovered while fixing https://github.com/pytest-dev/pytest-asyncio/issues/145.

@Zac-HD: Do you have any ideas?

**Comment by Zac-HD** [comment]
This might be as simple as reversing our iteration order, or we might need to do something a little fancier... 

We also need to add tests for cases where a function-scoped fixture is overridden with a safe scope, and vice-versa!

**Comment by simonfagerholm** [comment]
@Zac-HD: I did a quick test and this approach seems to work at least in this case:
```
for fx in fx_defs:
	fxture = item._request._get_active_fixturedef(fx.argname)
	if fxture.scope == "function":
		note_deprecation(
			"%s uses the %r fixture, but function-scoped fixtures "
			"should not be used with @given(...) tests, because "
			"fixtures are not reset between generated examples!"
			% (item.nodeid, fx.argname),
			since="2020-02-29",
		)
```
I used the internal _get_active_fixturedef to get the same fixture that pytest is using for this item. I figured this might be better than guessing pytests approach to order fixtures, but it might have other problems.

**Comment by simonfagerholm** [comment]
Maybe this is better to prevent initializing fixtures that are not being used:
```
if fx.argname in argnames:
    fxture = item._request._get_active_fixturedef(
        fx.argname)
    if fxture.scope == "function":
        note_deprecation(
            "%s uses the %r fixture, but function-scoped"
            " fixtures should not be used with @given(...)"
            " tests, because fixtures are not reset "
            "between generated examples!"
            % (item.nodeid, fx.argname),
            since="2020-02-29",
        )
```

**Comment by Zac-HD** [comment]
That looks great to me - want to open a PR?

**Comment by simonfagerholm** [comment]
@Zac-HD Sure, I will get on it

**Comment by Zac-HD** [comment]
Closed by #2376.
